[Li Tianze] ip

TZL0 opened the pull request:
# 1. Duke 
## 2. A Bullet List on Request
- Bullet
- Bullet
- Bullet

## 3. A Number List on Request
1. Number
2. Number
3. Number

## 4. A fenced code block with syntax highlighting
```Java
System.out.println("This is Duke.");
```

## 5. A Task List on Request
- [x] Increment Level-1 to Level-9.
- [ ] Increment Level-10
- [ ] Increment A-Varargs 

## 6. An emoji on Request :star_struck:

## 7.  A blockquote
> Hello world!
## 8. This is a hyperlink:
[Google](https://www.google.com)

## 9. This is an inline code:
Line line line line `code` line line line line.

## 10. Some text formattings:
### 1. Bold
**text formatting**
### 2. Italic
*text formatting*
### 3. Stikethrough
~~text formatting~~

Commit:
Add Gradle support

TZL0 committed:
Built MessagePrinter

TZL0 committed:
Level-1 completed

TZL0 committed:
Level 2 completed

TZL0 committed:
MessagePrinter now supports change of the horizontal line characters and horizontal line length

TZL0 committed:
class Action, Parser, Command, Task created and Level-3 completed

TZL0 committed:
Bug fixed. Now list() starts with a message

TZL0 committed:
Level_4 completed. Optimized structure of classes.

TZL0 committed:
Level-5 completed. Now duke is able to detect and warn errors. Parser is renamed to Compiler.

TZL0 committed:
Level-6 completed. Now Duke supports deleting tasks from list.

TZL0 committed:
Duke can deal with empty entry well

TZL0 committed:
Remove unnecessary rows

TZL0 committed:
Separate compile time duke exception and runtime duke exception.

TZL0 committed:
Add A-TextUiTesting

TZL0 committed:
Remove Add(). Task.markAsDone() and Task.markAsNotDone replace Task.setIsDone.

TZL0 committed:
Add TaskList, Savable, DukeException.ReadAttributeException.
Update Task.

TZL0 committed:
Add save(), read().
Change ATTRIBUTE_SEPARATOR: "|" -> "}".

TZL0 committed:
Support Real Dates and Times.

TZL0 committed:
Put implementation of commands into class Commands.

TZL0 committed:
Put implementation of commands into class Command.

TZL0 committed:
Level-7 completed

TZL0 committed:
Level-8 completed

TZL0 committed:
Merged Level 7 and 8

TZL0 committed:
Divide inner classes of Command, DukeException, Task in to separate classes.
Divide classes into packages.

TZL0 committed:
Change some Access Modifiers.

TZL0 committed:
Build Gradle

TZL0 committed:
Add JUnit Test. Add equals().

TZL0 committed:
Add equals().

TZL0 committed:
Add A-JUnit.

TZL0 committed:
Add A-Jar

TZL0 committed:
Add A-CodingStandard

TZL0 committed:
Add find().

TZL0 committed:
Bug on Level-9 fixed.

TZL0 committed:
Add JavaDoc for duke.command

TZL0 committed:
Add JavaDoc for Duke and Duke's Segments.

TZL0 committed:
Add JavaDoc to Exceptions.

TZL0 committed:
Merge branch 'branch-Level-9' into branch-A-JavaDoc

# Conflicts:
#	src/main/java/duke/TaskList.java
#	src/main/java/duke/command/Command.java

TZL0 committed:
Merge branch 'branch-A-CodingStandard' into branch-A-JavaDoc

# Conflicts:
#	src/main/java/duke/Storage.java
#	src/main/java/duke/TaskList.java
#	src/main/java/duke/command/Action.java
#	src/main/java/duke/command/AddCommand.java
#	src/main/java/duke/command/DeadlineCommand.java
#	src/main/java/duke/command/DeleteCommand.java
#	src/main/java/duke/command/EventCommand.java
#	src/main/java/duke/command/UnmarkCommand.java
#	src/main/java/duke/exception/DukeRuntimeException.java
#	src/main/java/duke/exception/ReadAttributeException.java
#	src/main/java/duke/exception/TaskNotFoundException.java

TZL0 committed:
Merge remote-tracking branch 'upstream/add-gradle-support'

# Conflicts:
#	build.gradle
#	gradle/wrapper/gradle-wrapper.jar
#	gradle/wrapper/gradle-wrapper.properties
#	gradlew
#	gradlew.bat

TZL0 committed:
Add A-CheckStyle.
Comply more to Code Standard.

TZL0 committed:
Package utilities into duke.util.

melissaharijanto commented on src/main/java/duke/util/TaskList.java:
```diff
@@ -0,0 +1,164 @@
+     * @param taskArrayList The given ArrayList of Tasks.
+     * @return The boolean whether the given ArrayList is equals to the internal TaskList.
+     */
+    protected boolean compareTaskList(ArrayList<Task> taskArrayList) {
```
perhaps the method name can be named with the is, was, has, etc. prefix?

melissaharijanto commented on src/main/java/duke/command/DeadlineCommand.java:
```diff
@@ -0,0 +1,89 @@
+    }
+
+    /**
+     * Execute the Command with given Duke Segments.
```
according to the javadoc rule, i think this should be 'executes'.

melissaharijanto commented on src/main/java/duke/command/DeleteCommand.java:
```diff
@@ -0,0 +1,85 @@
+    }
+
+    /**
+     * Return the index of Task to be deleted in the TaskList.
```
same javadoc issue; should be 'returns'

melissaharijanto commented on src/main/java/duke/command/DeleteCommand.java:
```diff
@@ -0,0 +1,85 @@
+    }
+
+    /**
+     * Execute the Command with given Duke Segments.
```
same javadoc issue as above.

melissaharijanto reviewed (commented):
Your code is really good! I have almost nothing to say to them + also kind of the reason why my comments may come as nitpicky (I'm so sorry 😭 ) but your implementation of Duke has given me insight on the various ways to implement this (love how you used a hashmap!) and your code is really OOP. Keep up the good work!

TZL0 committed:
Add Naive GUI without FXML

TZL0 committed:
Reorganize GUI with FXML

TZL0 committed:
Add MainWindow.boot()
Add Storage.write(Savable)

TZL0 committed:
Add A-CheckStyle

TZL0 committed:
Merge branch 'branch-A-CheckStyle' into branch-Level-10

TZL0 committed:
Add A-Varargs

yuanxi1 commented on src/main/java/duke/command/Action.java:
```diff
@@ -0,0 +1,137 @@
+
+    /**
+     * Returns the String representation of an Action.
+     * @param action The Action to convert.
```
Could add an empty line between description and parameter section. I notice the same issue in other places too.

yuanxi1 commented on src/main/java/duke/util/Parser.java:
```diff
@@ -0,0 +1,394 @@
+            indexOfName = Action.getString(action).length();
+            indexOfTime = getFirstIndexOfStr1InStr2(symbolEvent, s);
+            arg1 = s.substring(indexOfName, indexOfTime).trim();
+            arg2 = s.substring(indexOfTime + symbolEvent.length()).trim();
```
Is there any reason for defining a `getFirstIndexOfStr1InStr2` function for getting the second argument? I
 was thinking would it be simpler to use [`split`](https://docs.oracle.com/javase/7/docs/api/java/lang/String.html#split(java.lang.String,%20int))?

yuanxi1 commented on src/main/java/duke/exception/InvalidActionException.java:
```diff
@@ -0,0 +1,36 @@
+/**
+ * The DukeException when user enters an invalid action keyword.
+ */
+public class InvalidActionException extends CompileException {
```
I really like how you defined different types of duke exception sub-classes.

yuanxi1 commented on src/main/java/duke/command/Action.java:
```diff
@@ -0,0 +1,137 @@
+    /**
+     * The HashMap stores the standardFormat of an Action.
+     */
+    private static HashMap<Action, String> actionFormatMap = new HashMap<>();
```
I really like your use of enums and HashMap to map the command to action

yuanxi1 reviewed (commented):
Overall, I think you followed OOP principles very well and there are very few coding standard violations. Your project is well structured and I enjoyed reading your code :)

TZL0 committed:
Main Window: Add 'Shutting down...' when executing ExitCommand.

Make the duke GUI more realistic as a bot.

When the duke stops running, MainWindow.end() is called.

MainWindow is the JavaFx Class.

TZL0 committed:
MainWindow: add import Random

MainWindow.getNextSleepTime() need the Random Number generator.

Add import statement, add field random declaration.

Guarantee the duke works.

TZL0 committed:
Replace for-loop in Action.java by Stream.

TZL0 committed:
Rename boolean method so it sounds like boolean method.

TZL0 committed:
Reorganize Parser.java
To improve code quality.
Split parseCommand() into multiple methods.
Reduce method length.

TZL0 committed:
Add 's' to verbs in JavaDoc.

TZL0 committed:
Add assert.

TZL0 committed:
Merge pull request #4 from TZL0/branch-Bug-Fixed

Branch bug fixed

TZL0 committed:
Merge pull request #3 from TZL0/branch-A-CodeQuality

Branch a code quality

TZL0 committed:
Merge pull request #2 from TZL0/branch-A-Streams

Branch a streams

TZL0 committed:
Merge pull request #1 from TZL0/branch-A-Assertions

Branch a assertions

TZL0 committed:
Add C-DetectDuplicate

TZL0 committed:
Merge pull request #5 from TZL0/branch-C-DetectDuplicates

Add C-DetectDuplicate

TZL0 committed:
Create gradle.yml

TZL0 committed:
Fix Style

TZL0 committed:
Fix Magic Number and Magic String

TZL0 committed:
Add Task.clone()
Add TaskList.clone()
Add Duke.taskListHistory
Add getter and setter in Duke.java

TZL0 committed:
Change Command.execute(TaskList, MessagePrinter, Storage)
to Command.execute(Duke)

TZL0 committed:
Remove TaskList.loadFrom(), TaskList.getTaskArrayList()

TZL0 committed:
Fix Bug that shows incorrect number of loaded tasks.

TZL0 committed:
Complete C-Undo

TZL0 committed:
Update JavaDoc

TZL0 committed:
Remove Factory Methods of Command

TZL0 committed:
Remove Command.getFormat()

TZL0 committed:
Bug fixed

TZL0 committed:
Fix Code Quality

TZL0 committed:
Fix Bug of DukeException

TZL0 committed:
Add Ui.png

TZL0 committed:
Bug fixed (marked and unmarked tasks are equal).

TZL0 committed:
Update README.md

TZL0 committed:
Comply to code standard

TZL0 committed:
Update Text-ui-test

TZL0 committed:
Update build.gradle to generate correct jar

TZL0 committed:
Support Windows

TZL0 committed:
Update Tests

TZL0 committed:
Fix JavaFx.
Fix Coding Standard.
Fix Double Hello Message

TZL0 committed:
Fix 'Send' button when running without zulu

TZL0 committed:
Fix coding standard in MessagePrinter.
Fix JavaFx warning message.

TZL0 committed:
Improve terminal logging UI

TZL0 committed:
Fix Coding Standard

TZL0 committed:
Fix Coding Standard

TZL0 committed:
Remove unused plugin
